File: src/main/clojure/aqua/recommend/item_item_model.clj
(ns aqua.recommend.item-item-model
  (:require aqua.recommend.model-files))

(defn store-item-item [out item-item]
  (let [anime-indices (.animeRatedMap item-item)
        similar-ids (.similarAnimeId item-item)
        similar-scores (.similarAnimeScore item-item)
        writer (no.uib.cipr.matrix.io.MatrixVectorWriter. out)]
    (binding [*out* out]
      (println (.similarAnimeCount item-item))
      (println (count similar-ids))
      (println (count anime-indices))
      (.printArray writer anime-indices)
      (.printArray writer similar-ids)
      (.printArray writer similar-scores))))

(defn- load-item-item-v1 [in]
  (let [read-int (fn [] (Integer/valueOf (read-line)))
        reader (no.uib.cipr.matrix.io.MatrixVectorReader. in)]
    (binding [*in* in]
      (let [similar-count (read-int)
            scores-count (read-int)
            anime-indices (int-array (read-int))
            similar-ids (int-array scores-count)
            similar-scores (float-array scores-count)]
        (.readArray reader anime-indices)
        (.readArray reader similar-ids)
        (.readArray reader similar-scores)
        (aqua.recommend.ItemItemModel. anime-indices similar-count
                                        similar-ids
                                        similar-scores)))))

(defn load-item-item [path]
  (aqua.recommend.model-files/with-open-model path 1 in version
    (load-item-item-v1 in)))

(defn get-raw-anime-recommendations [user
                                     ^aqua.recommend.ItemItemModel model
                                     remove-known-anime]
  (let [ranked-anime (.findSimilarAnime model user)]
    (.sort ranked-anime aqua.recommend.ScoredAnimeId/SORT_SCORE)
    [[] (take 100 (remove-known-anime ranked-anime))]))

(defn get-raw-manga-recommendations [user
                                     ^aqua.recommend.ItemItemModel model
                                     remove-known-anime]
  (let [ranked-anime (.findSimilarManga model user)]
    (.sort ranked-anime aqua.recommend.ScoredAnimeId/SORT_SCORE)
    [[] (take 100 (remove-known-anime ranked-anime))]))

(defn get-anime-recommendations [user model remove-known-anime keep-airing-anime tagger]
  (let [[_ recommendations] (get-raw-anime-recommendations user model remove-known-anime)]
    [(tagger recommendations) []]))

(defn get-manga-recommendations [user model remove-known-anime keep-airing-anime tagger]
  (let [[_ recommendations] (get-raw-manga-recommendations user model remove-known-anime)]
    [(tagger recommendations) []]))

File: src/main/java/aqua/recommend/ItemItemModel.java
package aqua.recommend;

import com.carrotsearch.hppc.IntIntHashMap;
import com.carrotsearch.hppc.IntIntMap;
import com.carrotsearch.hppc.cursors.IntIntCursor;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ItemItemModel {
    private static final int ITEM_MIN = 10;
    private static final int ITEM_MAX = 20;

    public final IntIntMap animeIndexMap;
    public final int[] animeRatedMap;
    public final int similarAnimeCount;
    public final int[] similarAnimeId;
    public final float[] similarAnimeScore;

    public ItemItemModel(Map<Integer, Integer> animeIndexMap, int similarAnimeCount, int[] similarAnimeId,
            float[] similarAnimeScore) {
        this(HPPCUtils.convertMap(animeIndexMap), similarAnimeCount, similarAnimeId, similarAnimeScore);
    }

    public ItemItemModel(IntIntMap animeIndexMap, int similarAnimeCount, int[] similarAnimeId,
            float[] similarAnimeScore) {
        this.animeIndexMap = animeIndexMap;
        this.similarAnimeCount = similarAnimeCount;
        this.similarAnimeId = similarAnimeId;
        this.similarAnimeScore = similarAnimeScore;

        this.animeRatedMap = new int[animeIndexMap.size()];
        for (IntIntCursor entry : animeIndexMap)
            animeRatedMap[entry.value] = entry.key;
    }

    public ItemItemModel(int[] animeRatedMap, int similarAnimeCount, int[] similarAnimeId, float[] similarAnimeScore) {
        this.animeRatedMap = animeRatedMap;
        this.similarAnimeCount = similarAnimeCount;
        this.similarAnimeId = similarAnimeId;
        this.similarAnimeScore = similarAnimeScore;

        this.animeIndexMap = new IntIntHashMap();
        for (int i = 0; i < animeRatedMap.length; ++i)
            animeIndexMap.put(animeRatedMap[i], i);
    }

    public List<ScoredAnimeId> similarAnime(int animedbId) {
        Integer index = animeIndexMap.get(animedbId);
        if (index == null)
            return new ArrayList<>();
        List<ScoredAnimeId> result = new ArrayList<>(similarAnimeCount);

        for (int i = 0, currentIndex = index * similarAnimeCount; i < similarAnimeCount; ++i, ++currentIndex) {
            if (similarAnimeId[currentIndex] == 0)
                break;
            result.add(new ScoredAnimeId(similarAnimeId[currentIndex], similarAnimeScore[currentIndex]));
        }

        return result;
    }

    public List<ScoredAnime> findSimilarAnime(CFUser user) {
        List<ScoredAnimeId> sortedScores = new ArrayList<>(user.completedAndDroppedIds.length);
        for (int i = 0; i < user.completedAndDroppedIds.length; ++i) {
            sortedScores.add(new ScoredAnimeId(user.completedAndDroppedIds[i],
                user.completedAndDroppedRatingFloat(i)));
        }
        return sortScoredItems(sortedScores);
    }

    public List<ScoredAnime> findSimilarManga(CFUser user) {
        List<ScoredAnimeId> sortedScores = new ArrayList<>();
        for (CFRated item : user.inProgressAndDropped()) {
            sortedScores.add(new ScoredAnimeId(item.animedbId, user.normalizedRating(item)));
        }
        return sortScoredItems(sortedScores);
    }

    private List<ScoredAnime> sortScoredItems(List<ScoredAnimeId> sortedScores) {
        sortedScores.sort(ScoredAnimeId.SORT_SCORE);
        int maxUse = Math.min(ITEM_MAX, sortedScores.size());
        int minUse = Math.min(ITEM_MIN, sortedScores.size());
        List<ScoredAnimeId> picked = new ArrayList<>(maxUse);
        for (int i = 0; i < maxUse; ++i) {
            picked.add(sortedScores.get(sortedScores.size() - 1 - i));
        }
        Collections.shuffle(picked);
        int use = (int) (Math.random() * (maxUse + 1 - minUse)) + minUse;
        Map<Integer, ScoredAnimeId> recommendations = new HashMap<>();
        for (int i = 0; i < use; ++i) {
            ScoredAnimeId likedItem = picked.get(i);
            Integer likedAnimeIndex = animeIndexMap.get(likedItem.animedbId);
            if (likedAnimeIndex == null)
                continue;
            for (int j = 0, index = likedAnimeIndex * similarAnimeCount; j < similarAnimeCount; ++j, ++index) {
                int animedbId = similarAnimeId[index];
                if (animedbId == 0)
                    continue;
                float score = similarAnimeScore[index] * likedItem.score;
                ScoredAnimeId recommendation = recommendations.computeIfAbsent(animedbId,
                        id -> new ScoredAnimeId(id, 0.0f));
                recommendation.score += score;
            }
        }
        return new ArrayList<>(recommendations.values());
    }
}
